### PR TITLE
[JIT] Frozen Graph Conv-BN fusion

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50241 Re-enable no-op peepholes
* #50222 Add Post Freezing Optimizations, turn on by default in torch.jit.freeze
* #50221 Peephole Optimize out conv(x).dim(), which prevents BN fusion
* #50220 [JIT] Factor out peephole to own test file
* #50075 [JIT] Add Frozen Conv-> Add/Sub/Mul/Div fusion
* **#50074 [JIT] Frozen Graph Conv-BN fusion**

Adds Conv-BN fusion for models that have been frozen. I haven't explicitly tested perf yet but it should be equivalent to the results from @Chillee's PR [here](https://github.com/pytorch/pytorch/pull/476570) and [here](https://github.com/pytorch/pytorch/pull/47657#issuecomment-725752765). Click on the PR for details but it's a good speed up. 

 In a later PR in the stack I plan on making this optimization on by default as part of `torch.jit.freeze`. I will also in a later PR add a peephole so that there is not conv->batchnorm2d doesn't generate a conditional checking # dims.

Zino was working on freezing and left the team, so not really sure who should be reviewing this, but I dont care too much so long as I get a review 😛

Differential Revision: [D25856261](https://our.internmc.facebook.com/intern/diff/D25856261)

### CODE REVIEW

File: test/elias.py [hunk 1]
```diff
@@ -1,12 +0,0 @@
...
-@torch.jit.script
-def foo(x):
-    return x + x + x
-
-torch._C._jit_override_can_fuse_on_cpu(True)
-
-foo(torch.rand([2], requires_grad=False))
-foo(torch.rand([2], requires_grad=False))
-foo(torch.rand([2], requires_grad=False))
-print(torch.jit.last_executed_optimized_graph())
```
Comment (ZolotukhinM) [inline test/elias.py:12]: Wait, is it in the trunk?😃

File: torch/csrc/jit/passes/frozen_conv_folding.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+#include <torch/csrc/jit/passes/fold_conv_bn.h>
+#include <torch/torch.h>
+
+namespace torch {
+namespace jit {
+
+using Tensor = at::Tensor;
+
+bool nonConstantParameters(Node* n) {
+  for (size_t i = 1; i < n->inputs().size(); i++) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/frozen_conv_folding.cpp:16]: Why do we start from 1 here?

File: torch/csrc/jit/python/python_ir.cpp [hunk 3]
```diff
@@ -503,6 +503,17 @@ void initPythonIRBindings(PyObject* module_) {
...
+          "schema",
+          [](Node& n) {
+            std::stringstream ss;
+            if (auto sch = n.maybeSchema()) {
+              ss << n.schema();
+            } else {
+              ss << "(no schema)";
+            }
+            return ss.str();
+          })
```
Comment (ZolotukhinM) [inline torch/csrc/jit/python/python_ir.cpp:516]: Is it a leftover?

File: torch/csrc/jit/passes/frozen_conv_folding.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+          constant_as<bool>(n->namedInput("transposed")).value_or(true);
+      // dont handle transposed conv yet or not-constant transpose parameter
+      return !transposed_conv;
+    }
+    default:
+      return false;
+  }
+}
+
+void FoldFrozenConvBatchnorm(Block* b) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/frozen_conv_folding.cpp:41]: What is the difference between this pass and the one in `torch/csrc/jit/passes/fold_conv_bn.*`? Is it that this pass handles functional conv/bn sequences while the existing pass handles Conv-module/BN-module calls? If so, the name might be adjusted I think since it theoretically could occur in a non-frozen model too.

File: test/jit/test_freezing.py [hunk 5]
```diff
@@ -1331,3 +1332,59 @@ def forward(self, x: torch.Tensor, key: str) -> Any:
...
+            if tracing:
+                scripted_mod = torch.jit.trace(mod_eager, (inp))
+            else:
+                scripted_mod = torch.jit.script(mod_eager)
+
+            self.run_pass("inline", scripted_mod.graph)
+
+            # TODO: add optimization for conv size
+            # Currrently a size check in batch norm gives conv multiple uses
+            # and prevents fusion
```
Comment (ZolotukhinM) [inline test/jit/test_freezing.py:1373]: Does this mean that in practice this pass won't be able to do anything yet?

**Review by ZolotukhinM — APPROVED** [review]
Overall looks good, some comments are inline!

File: torch/csrc/jit/python/python_ir.cpp (hunk 3 above)
Comment (eellison) [reply torch/csrc/jit/python/python_ir.cpp:516]: No, I added this intentionally

File: torch/csrc/jit/passes/frozen_conv_folding.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/passes/frozen_conv_folding.cpp:16]: It was as of yesterday 😳

File: torch/csrc/jit/passes/frozen_conv_folding.cpp (hunk 4 above)
Comment (eellison) [reply torch/csrc/jit/passes/frozen_conv_folding.cpp:41]: The other pass handles modules. Basically, the other pass was written pre-freezing, I'm not sure there's any reason for it to exist anymore

File: test/jit/test_freezing.py (hunk 5 above)
Comment (eellison) [reply test/jit/test_freezing.py:1373]: Yes, for scripted models. I'm going to address that limitation later on in the stack

File: torch/csrc/jit/passes/frozen_conv_folding.cpp (hunk 4 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/passes/frozen_conv_folding.cpp:41]: Originally it was written for graph-mode quantization, which is replaced with an FX version now anyway. I think we should get rid of it then (cc @jerryzh168 ).

File: torch/csrc/jit/passes/frozen_conv_folding.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/passes/frozen_conv_folding.cpp:16]: oop, meant for the `test/elias.py` comment. We start from 1 because the input to conv isn't known, but we do need to know all other values (the conv/batchnorm parameeters)

File: torch/csrc/jit/passes/frozen_conv_folding.cpp (hunk 4 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/frozen_conv_folding.cpp:41]: this is added by @kimishpatel I think.

there are still people using torchscript graph mode quantization I think so we can't remove it right now.

Comment (eellison) [reply torch/csrc/jit/passes/frozen_conv_folding.cpp:41]: Doesn't graph mode quantization using freezing though? why can't we run fusion post - freezing?

Comment (jerryzh168) [reply torch/csrc/jit/passes/frozen_conv_folding.cpp:41]: process for quantization is: fusion - prepare - convert - freeze
fusion needs to happen before freeze because it affects the weight of conv which will be observed in prepare and quantized in convert

Comment (kimishpatel) [reply torch/csrc/jit/passes/frozen_conv_folding.cpp:41]: As Jerry said the original conv+bn fusion cannot work post freezing. Plus since that one works at module levels we cannot quite make it work post freezing which inlines the graph. Ideally we add graph model fusion which works directly on graphs instead of modules, which this PR seems to be adding. Right now module based fusion is used in optimize_for_mobile. We can potentially move it after freezing. If this pass is for graph based fusion then is it necessary that it works only on frozen graph? Can it be made to work on non-frozen graphs too? That might be useful.

Comment (eellison) [reply torch/csrc/jit/passes/frozen_conv_folding.cpp:41]: There's not a great way to make this work on graphs that aren't frozen. You would have to remove the batchnorm weights from the parent module which isn't in general a valid transformation unless user-specified. That would also add a third conv + bn pass.
